### ISSUE TITLE
Saving a model causes the model's behavior to change

### ISSUE THREAD

**Issue by mmroden** [JohnLangford/vowpal_wabbit#641]
If I run the following code in a vagrant set up from this test project: https://github.com/mmroden/vw_daemon_tests

The following script, run from the /vagrant/data directory, causes output1.txt to not match output2.txt.  That is, saving a model causes the behavior of the system to change, and I would expect that saving would not cause any change.

    killall vw
    vw --daemon --port 26542 --holdout_off --loss_function logistic --learning_rate 5.0 --decay_learning_rate 0.95 --power_t 0 --initial_t0 -f /vagrant/model_test.vw --save_resume --pid_file /tmp/vw.pid --l11e-8 --l2 1e-8 --num_children 1
    nc localhost 26542 < train-examples.txt
    echo "save" | nc localhost 26542
    nc localhost 26542 < additional-examples.txt > output1.txt
    killall vw
    rm /vagrant/model_test.vw
    vw --daemon --port 26542 --holdout_off --loss_function logistic --learning_rate 5.0 --decay_learning_rate 0.95 --power_t 0 --initial_t0 -f /vagrant/model_test.vw --save_resume --pid_file /tmp/vw.pid --l11e-8 --l2 1e-8 --num_children 1
    nc localhost 26542 < train-examples.txt
    nc localhost 26542 < additional-examples.txt > output2.txt
    diff output1.txt output2.txt

I would expect those output files would not be different at all.

**Comment by mmroden** [comment]
Also, if I change the vw invocation to look like:

    vw -f /vagrant/model_test.vw --save_resume --daemon --port 26542 --num_children 1

then saving the model does not affect the performance of the system, ie, output1.txt is the same as output2.txt.

**Comment by JohnLangford** [comment]
I think this was addressed on the mailing list, so closing.

**Comment by mmroden** [comment]
Can you please tell me where on the mailing list?  Because it really looks like saving a model causes corruption.

**Comment by JohnLangford** [comment]
I'm confused by the below.  W.r.t. the original issue, turn off 
--decay_learning_rate.  This is designed for multi-pass learning which 
does not make sense in the daemon context.

-John

**Comment by mmroden** [comment]
But even so, if it's in the daemon context and the flag doesn't make sense, shouldn't the program ignore the flag?  And even if it doesn't, why should saving cause the behavior to change?  I would still expect file outputs to remain the same regardless of whether or not saving is done.

**Comment by arielf** [comment]
@mmroden   Two thoughts:

1) Could you provide a reproducible example with a much smaller data file?  train-examples.txt is very big and if there's a bug here, it should be reproducible and easier to fix, with a much smaller data-set.
2) Could you verify that `killall vw` actually kills all the vw processes in your env?  I suspect some process that is blocked on input doesn't actually get killed so it continues to update the model after the point you may think it doesn't.

**Comment by mmroden** [comment]
@arielf I made a pretty complete daemon testing environment here:

https://github.com/mmroden/vw_daemon_tests

This project starts a vw instance using ubuntu inside of a local virtual machine (or a remote one, if you stand up your vagrant that way) and then interacts with it remotely rather than inside of it.  Once you sync the project, you should be able to stand it up via

    make check

but it takes some time.  I'll take a look at getting you a smaller sample file later today.

**Comment by mmroden** [comment]
OK, I've uploaded the project with another explicit test of this case, but it looks like it's working.  So yeah, let's leave this closed, and if I can replicate the issue, then I'll open a separate issue.  The regularization tests are still failing in the project, but I'm checking over those as well to make sure that they're doing the right thing.
